### PR TITLE
Add initial MaxStack calculation

### LINKED ISSUE #93419: [API Proposal]: Make OpCode.StackChange() public
### Background and motivation

[OpCodes](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/OpCodes.cs#L258) represents MS IL instructions that used for emitting IL with [ILGenerator](https://learn.microsoft.com/dotnet/api/system.reflection.emit.ilgenerator). An `OpCode` instruction may pop value(s) from stack as an operand of the instruction and also may push value(s) into the stack as a result.  This stack change is needed for calculating [MaxStack](https://learn.microsoft.com/dotnet/api/system.reflection.methodbody.maxstacksize) value that is essential for populating method body. I could use the   `StackBehaviourPop` and `StackBehaviourPush` properties of `OpCode` to calculate this, but there is an internal method that gives this info directly:
https://github.com/dotnet/runtime/blob/946c5245aea149d9ece53fd0debc18328c29083b/src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/Opcode.cs#L48-L49
Making this public will be useful for calculating `MaxStack` faster
Related to https://github.com/dotnet/runtime/pull/93244/files#r1353013060

### API Proposal

```diff
namespace System.Reflection.Emit

public readonly partial struct OpCode : System.IEquatable<System.Reflection.Emit.OpCode>
{
    ...
    public StackBehaviour StackBehaviourPop { get }
    public StackBehaviour StackBehaviourPush { get }
    ...
+   public int EvaluationStackDelta { get }
}
```
### Alternative Designs

Other naming options: `EvaluationStackDelta`, `StackSizeDelta`, `StackTransitionDelta` or `StackDepthDelta` as it is showing the difference of evaluation stack transition.

### API Usage

```csharp
    private void UpdateStackSize(OpCode opCode)
    {
        _currentStack += opCode.EvaluationStackDelta;
        _maxStackSize = Math.Max(_maxStackSize, _currentStack);
    }
```

### PR DESCRIPTION
**Pull request by buyaa-n**
This is initial MaxStack calculation for simple (no branching) instructions, basically `UpdateStackSize(OpCode opCode)` will be called for each `OpCode` emit and updates the `_currentStack` based on its `StackBehaviourPush`, `StackBehaviourPop` values.
```cs
        private void UpdateStackSize(OpCode opCode)
        {
            _currentStack += GetStackChangeFor(opCode.StackBehaviourPush) + GetStackChangeFor(opCode.StackBehaviourPop);

            if (_currentStack > _maxStackSize)
            {
                _maxStackSize = _currentStack;
            }
        }
```
This calculation will be updated as needed when branching IL.Emit overloads implemented. 

As I cannot use the internal [OpCode.StackChange()](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/Opcode.cs#L48-L49) method created `int GetStackChangeFor(StackBehaviour stackBehaviour)` method

Contributes to https://github.com/dotnet/runtime/issues/92975
Fixes https://github.com/dotnet/runtime/issues/93419

### CODE REVIEW

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs:103]:
```suggestion
        public override void Emit(OpCode opcode) => EmitOpcode(opcode);
```

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs:167]: This condition will handle all negative numbers too, is that what we want? I think an exception or some other other mode would be more appropriate.

Comment (buyaa-n) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs:167]: I thought when casting to `uint` negative values would become positive large number something that > 3, no?

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs:161]: Yes, that is true. My statement was more related to the notion of all unknown values being converted to `Ldarg_3`. I think `Ldarg_0` or an exception being the failure mode makes more sense.

Comment (buyaa-n) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs:161]: I am OK with defaulting to `Ldarg_0`, though this is how the existing CoreLib `ILGenerator` does, maybe we should keep it as is for compat, what you think?

Comment (AaronRobinsonMSFT) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs:161]: Boo. Your instinct here is correct - keep the same behavior.

File: src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveILGeneratorTests.cs
Comment (buyaa-n) [inline src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveILGeneratorTests.cs:248]:
```suggestion
}
```

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs
Comment (stephentoub) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs:48]: I'm finding the formatting here difficult to read. Can you at least add a blank line after each line that has a => in it, in order to separate the groupings?

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs
Comment (buyaa-n) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs:76]:
```suggestion
                or StackBehaviour.Push0 => 0,
                
                StackBehaviour.Pop1
                or StackBehaviour.Popi
                or StackBehaviour.Popref
                or StackBehaviour.Varpop => -1,
                
                StackBehaviour.Pop1_pop1
                or StackBehaviour.Popi_pop1
                or StackBehaviour.Popi_popi
                or StackBehaviour.Popi_popi8
                or StackBehaviour.Popi_popr4
                or StackBehaviour.Popi_popr8
                or StackBehaviour.Popref_pop1
                or StackBehaviour.Popref_popi => -2,
                
                StackBehaviour.Popi_popi_popi
                or StackBehaviour.Popref_popi_pop1
                or StackBehaviour.Popref_popi_popi
                or StackBehaviour.Popref_popi_popi8
                or StackBehaviour.Popref_popi_popr4
                or StackBehaviour.Popref_popi_popr8
                or StackBehaviour.Popref_popi_popref => -3,
                
                StackBehaviour.Push1
                or StackBehaviour.Pushi
                or StackBehaviour.Pushi8
                or StackBehaviour.Pushr4
                or StackBehaviour.Pushr8
                or StackBehaviour.Pushref
                or StackBehaviour.Varpush => 1,
                
                StackBehaviour.Push1_push1 => 2,
                
                _ => throw new InvalidOperationException()
```

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs
Comment (buyaa-n) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs:44]: If we make the below API public, we don't need this `GetStackChangeFor` method and calculating the `StackBehaviourPush`, `StackBehaviourPop` diff with  `GetStackChangeFor(opCode.StackBehaviourPush) + GetStackChangeFor(opCode.StackBehaviourPop)` https://github.com/dotnet/runtime/blob/715293eeeddf76593679fba7ab8cfd451980225b/src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/Opcode.cs#L48-L49

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs
Comment (buyaa-n) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs:82]: Oops spacing
```suggestion

                StackBehaviour.Pop1
                or StackBehaviour.Popi
                or StackBehaviour.Popref
                or StackBehaviour.Varpop => -1,

                StackBehaviour.Pop1_pop1
                or StackBehaviour.Popi_pop1
                or StackBehaviour.Popi_popi
                or StackBehaviour.Popi_popi8
                or StackBehaviour.Popi_popr4
                or StackBehaviour.Popi_popr8
                or StackBehaviour.Popref_pop1
                or StackBehaviour.Popref_popi => -2,

                StackBehaviour.Popi_popi_popi
                or StackBehaviour.Popref_popi_pop1
                or StackBehaviour.Popref_popi_popi
                or StackBehaviour.Popref_popi_popi8
                or StackBehaviour.Popref_popi_popr4
                or StackBehaviour.Popref_popi_popr8
                or StackBehaviour.Popref_popi_popref => -3,

                StackBehaviour.Push1
                or StackBehaviour.Pushi
                or StackBehaviour.Pushi8
                or StackBehaviour.Pushr4
                or StackBehaviour.Pushr8
                or StackBehaviour.Pushref
                or StackBehaviour.Varpush => 1,

                StackBehaviour.Push1_push1 => 2,

                _ => throw new InvalidOperationException()
```

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs
Comment (buyaa-n) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs]: As per offline discussion created [API proposal](https://github.com/dotnet/runtime/issues/93419) for making this public, updated the PR accordingly

File: src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveILGeneratorTests.cs
Comment (steveharter) [inline src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveILGeneratorTests.cs:237]: Having an aggregate comparison for stack size like this is fine, but may not detect bugs at an opcode level since, for example, an off-by-one in one case may be offset by an opposite off-by-negative-one case. Should we check the current stack size after each Emit(opcode) by using reflection to investigate the current stack size in the generator (`m_curDepth`) or is this redundant in general due to other tests?

**Review by steveharter — APPROVED** [review]
LGTM pending API approval.

**Comment by steveharter** [comment]
Note there are some unaccounted test failures. Perhaps rebase\merge to latest since I see some recent JIT fixes went in.

File: src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveILGeneratorTests.cs
Comment (buyaa-n) [reply src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveILGeneratorTests.cs:237]: > Should we check the current stack size after each Emit(opcode) by using reflection to investigate the current stack size in the generator (m_curDepth) or is this redundant in general due to other tests?

We can check stack size after each Emit(opcode), but might be redundant, especially in this PR only simple Emit operations covered that changes the stack linearly. My plan was keep checking the `MaxStackSize` for next PRs that add more complicated Emit overloads. It might be good to test current stack size for those Emit overloads that change the stack differently.

File: src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/Opcode.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/Opcode.cs:48]: This needs doc comment.

File: src/libraries/System.Reflection.Primitives/ref/System.Reflection.Primitives.cs
Comment (jkotas) [inline src/libraries/System.Reflection.Primitives/ref/System.Reflection.Primitives.cs:36]: Is this order produced by the ref assembly generator?
 
If you have not done it already, you may want to regenerate this ref assembly source (https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/updating-ref-source.md).

File: src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/Opcode.cs
Comment (buyaa-n) [reply src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/Opcode.cs:48]: Yes, thank you!

Comment (buyaa-n) [reply src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/Opcode.cs:54]: Added

File: src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/Opcode.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/Opcode.cs:52]: This needs to mention what this API returns when this is unknown (e.g. for `call` instruction)

**Comment by jkotas** [comment]
Add test for the new API under https://github.com/dotnet/runtime/blob/fe043b489f69a888ad2efd97fbc177c1f25e541d/src/libraries/System.Reflection.Emit/tests/OpCodesTests.cs

File: src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/Opcode.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/Opcode.cs:53]: You are right that the actual value is not fixed, but this method is going to return 0 in this case. It means that 0 returned by this method means either zero delta or unknown delta.

Would it make more sense to throw an exception for unknown delta instead of returning 0?

Comment (buyaa-n) [reply src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/Opcode.cs:53]: > Would it make more sense to throw an exception for unknown delta instead of returning 0?

That would change the existing `RuntimeILGenerator` behavior, I think I should note that in that case user should not rely on this value for calculating `StackSize` nor for calculating `MaxStack`.

Comment (buyaa-n) [reply src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/Opcode.cs:53]: > That would change the existing `RuntimeILGenerator` behavior 

Actually, it's not have to affect `RuntimeILGenerator` as we can handle the throwing in the new ILGenerator implementation, though not sure such special casing is worth, don't you think adding a `NOTE` in the doc could be enough?

Comment (buyaa-n) [reply src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/Opcode.cs:53]: > Actually, it's not have to affect `RuntimeILGenerator` as we can handle the throwing in the new ILGenerator implementation, 

I take it back, handling the throwing in the new ILGenerator doesn't make sense 😄

File: src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/Opcode.cs
Comment (buyaa-n) [inline src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/Opcode.cs:54]: How is this look?
```suggestion
        /// For some IL instructions like <see cref="OpCodes.Call"/> this value is not fixed and depends on the called reference signature. 
        /// In this case you should not rely on this value for calculating stack size and/or max stack, instead need to evaluate the reference signature.
        /// For example, in case it is calling a method, need to evaluate the method signature, the push count depends on the returning value,
        /// the pop count depends on how many parameters passed.
```

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/Opcode.cs:54]: In the first sentence, "this value" means "The difference between how many elements are popped from the stack and how many are pushed onto the stack".

In the second sentence, "this value" means "the value returned by this property".

Avoid saying "this value" for clarity?

Also, I think we should document what the function returns when the value is not fixed.

**Comment by buyaa-n** [comment]
Failures unrelated and already reported in https://github.com/dotnet/runtime/issues/88102, https://github.com/dotnet/runtime/issues/88027

**Pull request merged**
